components/support/nimbus-fml/src/defaults/hasher.rs

```diff
@@ -0,0 +1,305 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::schema::TypeQuery;
+use crate::{
+    intermediate_representation::{FeatureDef, ObjectDef, PropDef, TypeRef},
+    schema::Sha256Hasher,
+};
+use serde_json::Value;
+use std::{
+    collections::{BTreeMap, BTreeSet, HashSet},
+    hash::{Hash, Hasher},
+};
+
+pub(crate) struct DefaultsHasher<'a> {
+    object_defs: &'a BTreeMap<String, ObjectDef>,
+}
+
+impl<'a> DefaultsHasher<'a> {
+    pub(crate) fn new(objs: &'a BTreeMap<String, ObjectDef>) -> Self {
+        Self { object_defs: objs }
+    }
+
+    pub(crate) fn hash(&self, feature_def: &FeatureDef) -> u64 {
+        let mut hasher: Sha256Hasher = Default::default();
+        feature_def.defaults_hash(&mut hasher);
+
+        let types = self.all_types(feature_def);
+
+        // We iterate through the object_defs because they are both
+        // ordered, and we want to maintain a stable ordering.
+        // By contrast, `types`, a HashSet, definitely does not have a stable ordering.
+        for (obj_nm, obj_def) in self.object_defs {
+            if types.contains(&TypeRef::Object(obj_nm.clone())) {
+                obj_def.defaults_hash(&mut hasher);
+            }
+        }
+
+        hasher.finish()
+    }
+
+    fn all_types(&self, feature_def: &FeatureDef) -> HashSet<TypeRef> {
+        let types = TypeQuery::new(self.object_defs);
+        types.all_types(feature_def)
+    }
+}
+
+trait DefaultsHash {
+    fn defaults_hash<H: Hasher>(&self, state: &mut H);
+}
+
+impl DefaultsHash for FeatureDef {
+    fn defaults_hash<H: Hasher>(&self, state: &mut H) {
+        self.props.defaults_hash(state);
+    }
+}
+
+impl DefaultsHash for Vec<PropDef> {
+    fn defaults_hash<H: Hasher>(&self, state: &mut H) {
+        let mut vec: Vec<&_> = Default::default();
+        for item in self {
+            vec.push(item);
+        }
```

Comment:
```suggestion
        let mut vec = self.iter().collect::<Vec<_>>();
```

This will do a single allocation for `vec` instead of growing asympototically to ~`self.len()`
Also we don't need the `&` in the type hint because Vec<T>::iter() return an iterator over &T